sound/soc/sdca/sdca_function_device.c

```diff
@@ -86,6 +86,33 @@ static void sdca_dev_unregister(struct sdca_dev *sdev)
 	auxiliary_device_uninit(&sdev->auxdev);
 }
 
+int sdca_dev_parse_functions(struct sdw_slave *slave)
+{
+	struct sdca_device_data *sdca_data = &slave->sdca_data;
+	struct device *dev = &slave->dev;
+	struct sdca_function_desc *func_desc;
+	struct sdca_function_data *func;
+	int ret;
+	int i;
+
+	for (i = 0; i < sdca_data->num_functions; i++) {
+		func_desc = &sdca_data->sdca_func[i];
+
+		func = devm_kzalloc(dev, sizeof(*func), GFP_KERNEL);
+		if (!func)
+			return -ENOMEM;
+
+		func_desc->function = func;
+
+		ret = sdca_parse_function(dev, func_desc->function_node, func_desc);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+EXPORT_SYMBOL_NS(sdca_dev_parse_functions, SND_SOC_SDCA);
```

Comment:
Yeah again commit messages are still very lacking in this chain, apologies for that. I thought this might be the primary point of discussion between us on this stuff.

Firstly to be clear my intention is very much to have subdevices for each function, just to have them all share a common register map.

There are few reasons this makes sense, primarily its a single hardware device and generally one wants a register map to correspond to the hardware device. Only a single thing can talk to the hardware at one time, so it should be locked as such. Especially so in this case where SDCA only mandates a single MBQ buffer for the whole device, rather than one per function, which means you need several actual bus transactions to be atomic and not interrupted by another function. If you were to add a register map for each function you would have to have a lock at the soundwire device level that each subdevice took before it accessed the register map which is rather messy and error prone.

Secondly there are situations where one might want to access registers, especially the function level registers, before registering the subdevices. Currently the way we have this laid out, as per my comment on your patch chain, we force the function-type control to be a disco constant, however it would be totally spec compliant for that to be a device register. Creating the register map at this level would allow us to access these values as registers regardless of if they are a device register or a disco constant. Although I still need to work through your quirks issues and see if we can avoid forcing the constants there.

In short I see a lot of problems with creating a register map for each function and the only really advantage one gets is that it is slightly harder for one function to accidentally access another functions registers, which a function shouldn't be doing anyway. If a function is accessing the wrong registers it is likely the system is already not working properly.

Reply:
The premise has always been that Functions are independent.

The fact that there are choke points at the device level such as interrupt handling and MBQ buffers should not be a reason to rejoin functions together.

The mutual exclusion for MBQ access is not hard to add, it's one lock at the device level that needs to be taken/released in regmap_sdw_mbq_write/read.  It's simpler IMHO than creating a giant regmap that includes all the SDCA controls, even if/when there's no function driver probed.

That said, if for some reason a function driver needs to access a vendor-specific register that's not in the SDCA space, then I have no idea what to do. Even for SDCA memories for e.g. firmware download the regmap directions are not clear to me.

Reply:
We are not joining functions together just adding a common mechanism for accessing the hardware. Sharing the register map gets you all those things for free. Passing extra locks down into the register map is much worse layering than having a single mechanism to access the hardware. What problems are we actually trying to avoid with the separate register maps? All it does is add complexity and opportunities for disaster.

Reply:
well in hindsight that's another layering violation. We should have implemented the multiple MBQ read/writes in the SoundWire core with the proper API and locking. We currently did all this at the regmap layer by piggy-backing on regmap locking, so now regmap is the answer to everything. When all you have is a hammer, etc.

I really don't think it would be a 'disaster' to keep registers in the scope in which they are defined, and to deal with the transport separately.

Another practical aspect is that it'll be rather unlikely that all functions in a multi-function device are used at the same time. So if you only want to use the mic function, it'd be handy to only have to deal with regmap for the mic part only. Do we really want to do a regmap_sync() for everything or just for what we need when resuming?

Reply:
Hm... will have a think about what MBQing at the SoundWire level would look like, although the MBQ stuff has always felt like a protocol that sits on top of SoundWire to me, which is why I chose the regmap layer to put it in. If anything since we added the function_status stuff into it rather than the retrying I was nervous it was a bit low in the hierarchy, since we are now assuming the function_status control exists, which granted it is mandatory but you know people do things. Directly assuming SDCA controls at the soundwire level feels like we are poking up through the layers quite a lot.

Syncing the registers is a fair point in favour of multiple register maps, it would be nice to sync less, although in most practical situations I am not sure we are talking about that many registers here but it is probably worth investigating.

Reply:
At the risk of starting another argument :-) I would argue it makes sense to only register the subdevices once the SoundWire device has shown up on the bus. This is what the cs42l43 example there does. This has many advantages meaning a lot of the hacks around waiting for the first time the device shows up on the bus (that we have in a lot of the current SoundWire devices) are no longer necessary. It also means you don't get madness like the soundcard being available to user-space but it not actually being able to do anything.

The init writes are also an interesting point, they feel very much like they should be modeled as a register_patch. The primary reason being they are likely to access the registers in a weird direct way (ie. either open coding the MBQ or not using the MBQ at all and directly addressing each byte). So it almost certainly makes sense to bypass the cache with those and just have them go straight to the hardware, as in a register patch.

It does raise an interesting question in my head of: can a function that has init writes, not require initialisation but still have lost the register values?

Reply:
There are pros and cons on the subdevice registration in the probe or update_status callback

probe:
- pros: all error handling is easier to do, fail big and early(tm). 
- cons: need to deal with the device being attached on the bus

update_status
- pros: no need to deal with additional signaling
-  cons: error handling is really messy.

The main reason why we used the regmap cache mode is so that we could use devm_ in the peripheral probe, that wouldn't work in the update_status callback.

But if you register a device, then devm_ is allowed in the probe, so it's only the error handling that's messy.

the init writes cannot use the MBQ, the format is a list of pairs of 32-bit address and 8 bit data. I guess you could encode the MBQ sequence but that would have to be split in N transactions. Deferred writes are not allowed either.

In theory according to the spec the Function_Needs_Initialization bit might be set at any time, but it's hard to think of a design where this would be required.